### ISSUE TITLE
[doc] user specified associated data API

### ISSUE THREAD

**Issue by alxiong** [RustCrypto/AEADs#224]
Currently [documentation](https://docs.rs/aead/0.3.2/aead/trait.Aead.html#tymethod.encrypt) on `trait aead::Aead` and specific instantiation of AEAD suites only cover how to directly encrypt a message with (by default) an empty associated data field,

> The Payload type can be used to provide Additional Associated Data (AAD) along with the message: this is an optional bytestring which is not encrypted, but is authenticated along with the message. Failure to pass the same AAD that was used during encryption will cause decryption to fail, which is useful if you would like to "bind" the ciphertext to some other identifier, like a digital signature key or other identifier.

If I construct my own `Payload` with msg and aad specified, then how to pass to the `encrypt()` function is not so clear to me, cuz Payload itself doesn't implement `Into<Payload>`. 

I could be misunderstanding things, but maybe some examples in the documentation would be helpful.
Thanks!

**Comment by tarcieri** [comment]
All types have a blanket impl of `impl<T> From<T> for <T>`:

https://docs.rs/aead/0.3.2/aead/struct.Payload.html#impl-From%3CT%3E

...and all types that `impl From` have a blanket impl for `Into`:

https://doc.rust-lang.org/std/convert/trait.Into.html#impl-Into%3CU%3E

I agree the documentation for `aead::Aead` could be better there.

**Comment by alxiong** [comment]
Ah I see, thanks for the clarification! 

Understood now, can close the issue. 👍🏻

**Comment by tarcieri** [comment]
I'll leave it open to track adding better documentation

**Comment by alxiong** [comment]
hmm... :thinking: the following gives me Err, not sure why

```rust
// message: &[u8], assoc_data: &u8, nonce and recipient_box is properly constructed.
let ct = recipient_box.encrypt(&nonce,
                               Payload { msg: &message,
                                         aad: &assoc_data });
```

**Comment by tarcieri** [comment]
`crypto_box` and the `xsalsa20poly1305` construction it's based on doesn't support AAD as an algorithm. Unfortunately it's unclear from the current API it's unsupported and you're getting an unhelpful error.

We have this tracking issue open for traits which better reflect constructions like `xsalsa20poly1305` which don't support AAD: https://github.com/RustCrypto/traits/issues/261

**Comment by alxiong** [comment]
>the xsalsa20poly1305 construction it's based on doesn't support AAD as an algorithm

so will the team add support in the near future? or this exclusion has some rationale behind?

**Comment by tarcieri** [comment]
The algorithm itself has no support for AAD. It's the only such algorithm in this repo which doesn't.

The similar `chacha20poly1305` construction supports it. We could potentially add support for it to `crypto_box`

**Comment by alxiong** [comment]
>The similar chacha20poly1305 construction supports it. We could potentially add support for it to crypto_box

that would be great

**Comment by tarcieri** [comment]
I think this was ultimately addressed in #225
